beetsplug/substitute.py

```diff
@@ -34,8 +35,9 @@ def tmpl_substitute(self, text):
         """Do the actual replacing."""
         if text:
             for pattern, replacement in self.substitute_rules:
-                if pattern.match(text.lower()):
-                    return replacement
+                new_text, subs_made = re.subn(pattern, replacement, text)
+                if subs_made > 0:
+                    return new_text
```

Comment:
Does this mean that only the very first matching pattern is applied?

Reply:
Yes, and this was the case previously as well. I'm not 100% certain it is the optimal behaviour, but it does have its advantages. One example is it makes it easy to exclude certain matches, by adding a no-op rule. For example, to prevent splitting a certain artist with `and` in its title:
```yaml
substitute:
- It's Jo and Danny: It's Jo and Danny
- ^(.*?)(,| &| and).*: \1
```

Reply:
Maybe I got it wrong, but it seems to me that this design will only ever apply the first substitution it matches?

Say, I want to replace all **ft**s with a comma, and only then split the artists:

```yaml
substitute:
  - \ +f(ea)?t(\.|uring)?: ", "
  - ^([^,]+),.*: \1
```

If I have `Artist ft. Someone`, I would expect to end up with `Artist`, but it seems like I would end up with `Artist, Someone` since only the first pattern would get applied.

Reply:
Your understanding is correct. I'm however not sure applying every matching rule in sequence would actually give that much more power. You could just as well write a single rule that transforms `Artist ft. Someone` to `Artist` directly.
```yaml
substitute:
  - (.+) +f(ea)?t(\.|uring)?.*: \1
```

I also do think it's worth being able to create exceptions by interrupting the flow, as with my `It's Jo and Danny` example above (where the group itself has 'and' in its name). I'm not sure how you'd handle that if every rule will be applied in sequence.

I personally think there's value to the simplicity of each case being handled by a single rule.

Reply:
Okay, I think that would also work for my purposes. It may get a bit messy to define rules a lot of exceptions (I currently only have around 5), but it's doable.

Your `(?i)`s do make me wonder if we should also leave it up to the rules to define case insensitivity. Currently it is enabled for all rules. This would however be a breaking change as the plugin hitherto has enabled it by default, affecting even the example configuration in the docs:
```yaml
substitute:
  .*jimi hendrix.*: Jimi Hendrix
```

Reply:
I don't see an issue with ignoring the case - to be honest my rules above would have looked slightly more organised if the patterns were case insensitive 😅